pandas/core/internals.py

```diff
@@ -4066,6 +4076,27 @@ def canonicalize(block):
         return all(block.equals(oblock)
                    for block, oblock in zip(self_blocks, other_blocks))
 
+    def unstack(self, unstacker):
+        """Return blockmanager with all blocks unstacked"""
+        dummy = unstacker(np.empty((0, 0)), value_columns=self.items)
+        new_columns = dummy.get_new_columns()
+        new_index = dummy.get_new_index()
+        new_blocks = []
+        mask_columns = np.zeros_like(new_columns, dtype=bool)
+
+        for blk in self.blocks:
+            bunstacker = unstacker(
```

Comment:
I would push almost all of the logic inside this loop *into* the ``_unstack(...)`` call on the block itself.